### ISSUE TITLE
Use keystone.{user,password} pillar-keys in glance exec-mod

### ISSUE THREAD

**Issue by 0xf10e** [saltstack/salt#26210]
As mentioned by @whytewolf in #26099 the pillar-keys for keystone-credentials should be identical for all OpenStack-related modules.

I've [got a branch](https://github.com/0xf10e/salt/tree/fix_glance_mod_cred) where I changed this but this causes strange 404s:

```python-traceback
operator@master:/srv/salt/saltstack.git% sudo salt \* saltutil.sync_all saltenv=base,openstack; sudo salt -C 'hw-ctrl or controller' glance.image_list
[...]
controller:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1146, in _thread_return
        return_data = func(*args, **kwargs)
      File "/var/cache/salt/minion/extmods/modules/glance.py", line 375, in image_list
        g_client = _auth(profile)
      File "/var/cache/salt/minion/extmods/modules/glance.py", line 148, in _auth
        return client.Client(api_version, endpoint, **kwargs)
      File "/usr/lib/python2.7/dist-packages/glanceclient/client.py", line 22, in Client
        return client_class(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/glanceclient/v2/client.py", line 38, in __init__
        image_model = self._get_image_model()
      File "/usr/lib/python2.7/dist-packages/glanceclient/v2/client.py", line 46, in _get_image_model
        schema = self.schemas.get('image')
      File "/usr/lib/python2.7/dist-packages/glanceclient/v2/schemas.py", line 84, in get
        _, raw_schema = self.http_client.json_request('GET', uri)
      File "/usr/lib/python2.7/dist-packages/glanceclient/common/http.py", line 266, in json_request
        resp, body_iter = self._http_request(url, method, **kwargs)
      File "/usr/lib/python2.7/dist-packages/glanceclient/common/http.py", line 249, in _http_request
        raise exc.from_response(resp, body_str)
    HTTPNotFound: {"error": {"message": "The resource could not be found.", "code": 404, "title": "Not Found"}} (HTTP 404)
```

Couldn't figure out how this happens yet but I think we should get this fixed before Beryllium gets released.

**Comment by whytewolf** [comment]
I agree and thank you @0xf10e

**Comment by jfindlay** [comment]
@0xf10e, @whytewolf, thanks for finding and working on these openstack issues.  We are working very hard on eliminating blockers for Beryllium so that we can release on time.  Do you think this can be resolved within the next few days or the next week?

**Comment by 0xf10e** [comment]
@jfindlay I'll try to fix this ASAP but I'm afraid it will take a few days of wading through `glanceclient` to figure out what causes the 404s.
401 I would understand when I can't use a session token acquired with the admin token for anything than keystone or something. The `_auth()` seems slightly different between the modules so one of them behaving strangely differently isn't too surprising.

More tomorrow.

**Comment by whytewolf** [comment]
okay. just testing. and your change does work for me. am not not sure why this is failing in your setup. 

unless you have the wrong settings for keystone.settings

**Comment by whytewolf** [comment]
a few things i can think of that might be skewing the results.

I'm on kilo.
I am not using multiple profiles yet. just a single set of keystone settings.

**Comment by 0xf10e** [comment]
I think I found my problem: `"GET /v2.0/v2/schemas/image HTTP/1.1" 404` in `/var/log/upstart/keystone.log`.
I'll take a closer look on this tomorrow.

**Comment by 0xf10e** [comment]
Ahhh, now I got it: `glanceclient` was using this endpoint for glance, not for keystone...
Guess I have to query keystone for the given service's endpoint then.

**Comment by whytewolf** [comment]
okay, I am getting an odd error with your current glance.py

```
alt 'ctrl*' glance.image_create name='CentOS 7' disk_format='raw' container_format='bare' location='http://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud.raw' visibility=public
ctrl.wolfcloud.bad4.us:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1037, in _thread_return
        return_data = func(*args, **kwargs)
      File "/var/cache/salt/minion/extmods/modules/glance.py", line 291, in image_create
        image = g_client.images.create(name=name, **kwargs)
      File "/usr/lib/python2.7/dist-packages/glanceclient/v1/images.py", line 288, in create
        data=image_data)
      File "/usr/lib/python2.7/dist-packages/glanceclient/common/http.py", line 262, in post
        return self._request('POST', url, **kwargs)
      File "/usr/lib/python2.7/dist-packages/glanceclient/common/http.py", line 214, in _request
        raise exc.CommunicationError(message=message)
    CommunicationError: Error finding address for http://10.40.0.2:929/v1/images: HTTPConnectionPool(host='10.40.0.2', port=929): Max retries exceeded with url: /v1/images (Caused by <class 'socket.error'>: [Errno 111] Connection refused)
```

I traced it down to this line
```
  g_endpoint_url = g_endpoint_url['internalurl'].strip('v2')
```
the endpoint I have returned for me does not have a v2 on it. so by trying to strip v2 from it it is striping the 2 off the end of the port. 

```
salt 'ctrl*' keystone.endpoint_get glance
ctrl.wolfcloud.bad4.us:
    ----------
    adminurl:
        http://10.40.0.2:9292
    id:
        ff41b79fd43c43c6ae8b7ef40bec951d
    internalurl:
        http://10.40.0.2:9292
    publicurl:
        http://10.40.0.2:9292
    region:
        LVNV
    service_id:
        6b8eb2cb02d145b68aed2d5616b4521d
```
not everyone forces glance to v2. such as mirantis.

**Comment by 0xf10e** [comment]
Should be fixed now: Replaced `g_endpoint_url['internalurl']).strip('v2')` with `re.sub('/v2', '', g_endpoint_url['internalurl'])`.

It's not really forcing to use v2 via the endpoint. It's more a "you'll use v2 anyway so I've already added it for you" that sometimes is done (and sometimes accidentally so via copy & paste from some howto).
But as I commented, on Icehouse I need to use v1 for creating the image. As soon as we got Juno or Kilo running I can try to figure out a way to only fall back to v1 when v2 doesn't work.

Thanks for testing, though!

**Comment by whytewolf** [comment]
actually reading the documentation for glanceclient. here is how it is listed. 

If the endpoint is unversioned and a value for version is not supplied, glanceclient falls back to v1

see http://docs.openstack.org/developer/python-glanceclient/#id37

so in essence. putting v2 in the endpoint is trying to force v2 all the time. and without it we are falling back to v1. [as long as you are glanceclient 0.15.0 and later]

which explains why i never had issues with the glance module before the changes. I have always used v1. [no version and no v2 on the url] this is true even in kilo.

**Comment by cachedout** [comment]
Closed via #26323
